### PR TITLE
Implementation of flush procedure added

### PR DESCRIPTION
**Pull request by skywojciechowskim**
Summary: Implementation of flush procedure
Type: Feature
Test Plan: Unittests, Component Tests, YT Conformance Tests
Jira: RIALTO-264

### CODE REVIEW

File: tests/componenttests/client/tests/mse/FlushTest.cpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
+/*
+ * If not stated otherwise in this file or this component's LICENSE file the
+ * following copyright and licenses apply:
+ *
+ * Copyright 2023 Sky UK
```
Comment (skyproudmanp) [inline tests/componenttests/client/tests/mse/FlushTest.cpp:5]: Year 2024

File: tests/unittests/media/server/gstplayer/genericPlayer/tasksTests/FlushTest.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
+/*
+ * If not stated otherwise in this file or this component's LICENSE file the
+ * following copyright and licenses apply:
+ *
+ * Copyright 2022 Sky UK
```
Comment (skyproudmanp) [inline tests/unittests/media/server/gstplayer/genericPlayer/tasksTests/FlushTest.cpp:5]: 2024

File: media/server/gstplayer/source/tasks/generic/Flush.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        source = sourceElem->second.appSrc;
+    }
+    if (!source)
+    {
+        RIALTO_SERVER_LOG_WARN("failed to flush - source is NULL");
+        return;
+    }
+
+    // Clear/invalidate old NeedDatas
+    if (MediaSourceType::AUDIO == m_type)
```
[the remark this answers is not in the document]
Comment (skywojciechowskim) [reply media/server/gstplayer/source/tasks/generic/Flush.cpp]: There was an early return for UNKNOWN, but good point, it wouldn't work for new types. Changed.

File: media/server/gstplayer/source/tasks/generic/Flush.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    GstFormat format{GST_FORMAT_UNDEFINED};
+    gint64 start{0};
+    gint64 stop{0};
+    if (!m_resetTime)
+    {
+        m_gstWrapper->gstElementQueryPosition(m_context.pipeline, GST_FORMAT_TIME, &position);
+        GstQuery *query{m_gstWrapper->gstQueryNewSegment(GST_FORMAT_TIME)};
+        if (m_gstWrapper->gstElementQuery(m_context.pipeline, query))
+        {
+            m_gstWrapper->gstQueryParseSegment(query, &rate, &format, &start, &stop);
```
[the remark this answers is not in the document]
Comment (skywojciechowskim) [reply media/server/gstplayer/source/tasks/generic/Flush.cpp:103]: changed.

File: media/server/gstplayer/source/tasks/generic/Flush.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    {
+        RIALTO_SERVER_LOG_WARN("failed to send flush-start event");
+    }
+
+    // Update segment if we don't reset time
+    if (!m_resetTime)
+    {
+        GstSegment *segment{m_gstWrapper->gstSegmentNew()};
+        m_gstWrapper->gstSegmentInit(segment, GST_FORMAT_TIME);
+        m_gstWrapper->gstSegmentDoSeek(segment, rate, GST_FORMAT_TIME, GST_SEEK_FLAG_NONE, GST_SEEK_TYPE_SET, position,
```
[the remark this answers is not in the document]
Comment (skywojciechowskim) [reply media/server/gstplayer/source/tasks/generic/Flush.cpp:120]: I changed init values for rate and stop to default ones in GstSegment. If query fails, segment will be initialized with standard values.

File: tests/componenttests/client/tests/mse/FlushTest.cpp (hunk 1 above)
Comment (skywojciechowskim) [reply tests/componenttests/client/tests/mse/FlushTest.cpp]: changed.

File: tests/unittests/media/server/gstplayer/genericPlayer/tasksTests/FlushTest.cpp (hunk 2 above)
Comment (skywojciechowskim) [reply tests/unittests/media/server/gstplayer/genericPlayer/tasksTests/FlushTest.cpp]: changed.

File: proto/mediapipelinemodule.proto [hunk 6]
```diff
@@ -427,7 +444,7 @@ message PlaybackStateChangeEvent {
         PLAYING = 2;        ///< The backend player is playing media.
         PAUSED = 3;         ///< The backend player is paused.
         SEEKING = 4;        ///< The backend player is seeking a new playback position.
-        FLUSHED = 5;        ///< The backend player has flushed the media data.
+        SEEK_DONE = 5;      ///< The backend player has finished seek.
```
[the remark this answers is not in the document]
Comment (skywojciechowskim) [reply proto/mediapipelinemodule.proto:447]: Adding new param will break seek functionality. If newer server sends SEEK_DONE with new int (let's say 9) and client is not ready for that, seek will always fail. According to Stack Overflow, we can rename enum params, if we use binary protobuf format: https://stackoverflow.com/questions/50857940/renaming-enum-value-in-protobuf-backward-compatibility

**Review by skyproudmanp — APPROVED** [review]

**Pull request merged**